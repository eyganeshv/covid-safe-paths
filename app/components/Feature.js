import React from 'react';

import { Flag } from '../helpers/flags';

/**
 * Small wrapper around `<Flag />` which makes the default case easier:
 *
 * Usage:
 *
 * ```
 * <Feature name="google_import" fallback={() => <hr />}>
 *   <FeatureUi />
 * </Feature>
 * ```
 *
 * @param {{
 *   name: string;
 *   fallback?: () => import('react').ReactNode;
 *   children: import('react').ReactNode;
 * }} param0
 */
export const Feature = ({ name, fallback, children }) => {
  const keyPath = name.split('.');

  return (
    <Flag name={keyPath} render={() => children} fallbackRender={fallback} />
  );
};
