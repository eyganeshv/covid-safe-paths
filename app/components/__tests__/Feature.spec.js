import { render } from '@testing-library/react-native';
import React from 'react';
import { Text } from 'react-native';

import { FlagsProvider } from '../../helpers/flags';
import { Feature } from '../Feature';

it('renders feature if the flag is enabled', () => {
  const { asJSON } = render(
    <FlagsProvider flags={{ feature1: true }}>
      <Feature name='feature1'>
        <Text>feature1</Text>
      </Feature>
    </FlagsProvider>,
  );

  expect(asJSON()).toMatchSnapshot();
});

it('allows dotted notation in the feature name', () => {
  const { asJSON } = render(
    <FlagsProvider flags={{ feature1: { child: true } }}>
      <Feature name='feature1.child'>
        <Text>feature1.child</Text>
      </Feature>
    </FlagsProvider>,
  );

  expect(asJSON()).toMatchSnapshot();
});

it('does not render if dotted notation key path is falsy', () => {
  const { asJSON } = render(
    <FlagsProvider flags={{ feature1: {} }}>
      <Feature name='feature1.child'>
        <Text>feature1.child</Text>
      </Feature>
    </FlagsProvider>,
  );

  expect(asJSON()).toMatchSnapshot();
});

it('omits feature if the flag is disabled', () => {
  const { asJSON } = render(
    <FlagsProvider flags={{ feature1: false }}>
      <Feature name='feature1'>
        <Text>feature1</Text>
      </Feature>
    </FlagsProvider>,
  );

  expect(asJSON()).toMatchSnapshot();
});

it('omits feature if the flag is missing', () => {
  const { asJSON } = render(
    <FlagsProvider flags={{}}>
      <Feature name='feature1'>
        <Text>feature1</Text>
      </Feature>
    </FlagsProvider>,
  );

  expect(asJSON()).toMatchSnapshot();
});

it('renders the fallback instead, if the flag is disabled/omitted', () => {
  const { asJSON } = render(
    <FlagsProvider flags={{ feature1: false }}>
      <Feature name='feature1' fallback={() => <Text>Old UI</Text>}>
        <Text>feature1</Text>
      </Feature>
    </FlagsProvider>,
  );

  expect(asJSON()).toMatchSnapshot();
});
