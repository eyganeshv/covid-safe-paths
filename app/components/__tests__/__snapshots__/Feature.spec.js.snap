// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allows dotted notation in the feature name 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text>
    feature1.child
  </Text>
</View>
`;

exports[`does not render if dotted notation key path is falsy 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;

exports[`omits feature if the flag is disabled 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;

exports[`omits feature if the flag is missing 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;

exports[`renders feature if the flag is enabled 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text>
    feature1
  </Text>
</View>
`;

exports[`renders the fallback instead, if the flag is disabled/omitted 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text>
    Old UI
  </Text>
</View>
`;
